Modernize ReplayGain PreAmp Settings

complexlogic opened the pull request:
## Description
This PR makes the following changes:
- Change the ReplayGain PreAmp settings to be centered at 0 dB instead of 89 dB
- Use a slider for the GUI control
- Allow increments of 0.2 dB instead of 1 dB

## Motivation and context
The current Kodi ReplayGain PreAmp settings assume a reference level of 89 dB SPL/absolute (equivalent to -14 dB relative to digital full scale). However, the vast majority of modern ReplayGain scanners normalize to -18 dBFS by default, including:
- [foobar2000](https://www.foobar2000.org/)
- [r128gain](https://github.com/desbma/r128gain)
- [loudgain](https://github.com/Moonbase59/loudgain)
- [rsgain](https://github.com/complexlogic/rsgain)

...and others. The 89 dB SPL reference level is outdated and no longer in widespread use.

Kodi is the only music player I use that has the PreAmp control centered at 89 dB. Most modern ReplayGain-compatible players use a slider centered at 0 dB for PreAmp, similar to what I am proposing. Moreover, since the setting no longer depends on a particular reference level, it will always be technically correct regardless of which level the audio files were normalized to.
 
## How has this been tested?
I enabled logging, checked that the ReplayGain values outputted by the [ReplayGain log statement](https://github.com/xbmc/xbmc/blob/master/xbmc/cores/paplayer/AudioDecoder.cpp#L375-L378) are identical with the equivalent before/after settings, e.g. -5 instead of 84.

## What is the effect on users?
The new setting is much more straightforward and will be less confusing for users because it is similar to what is found in other players. Users that do have existing PreAmp settings will need to re-enter them because this change is not backwards compatible.

## Screenshots (if appropriate):
![screen19](https://user-images.githubusercontent.com/95071366/193733166-44e07720-5eb4-4d9c-9af4-2b2227a64019.png)


## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [x] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [x] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

CastagnaIT commented on xbmc/application/ApplicationVolumeHandling.cpp:
```diff
@@ -139,9 +139,9 @@ void CApplicationVolumeHandling::CacheReplayGainSettings(const CSettings& settin
   // initialize m_replayGainSettings
   m_replayGainSettings.iType = settings.GetInt(CSettings::SETTING_MUSICPLAYER_REPLAYGAINTYPE);
-  m_replayGainSettings.iPreAmp = settings.GetInt(CSettings::SETTING_MUSICPLAYER_REPLAYGAINPREAMP);
+  m_replayGainSettings.iPreAmp = static_cast<float>(settings.GetNumber(CSettings::SETTING_MUSICPLAYER_REPLAYGAINPREAMP));
```
GetNumber is double and you cast to float, but the iPreAmp is integer
not so good here

CastagnaIT commented on xbmc/application/ApplicationVolumeHandling.cpp:
```diff
@@ -139,9 +139,9 @@ void CApplicationVolumeHandling::CacheReplayGainSettings(const CSettings& settin
+  m_replayGainSettings.iPreAmp = static_cast<float>(settings.GetNumber(CSettings::SETTING_MUSICPLAYER_REPLAYGAINPREAMP));
   m_replayGainSettings.iNoGainPreAmp =
-      settings.GetInt(CSettings::SETTING_MUSICPLAYER_REPLAYGAINNOGAINPREAMP);
+      static_cast<float>(settings.GetNumber(CSettings::SETTING_MUSICPLAYER_REPLAYGAINNOGAINPREAMP));
```
same here

CastagnaIT commented on xbmc/application/ApplicationVolumeHandling.h:
```diff
@@ -24,8 +24,8 @@ class CApplicationVolumeHandling
   {
-    int iPreAmp;
-    int iNoGainPreAmp;
+    float iPreAmp;
```
i saw now,
suggest to rename `iPreAmp` to `m_preAmp` to prevent confusions
also for `iNoGainPreAmp`

complexlogic replied:
I changed all the members of the `ReplayGainSettings` struct so they are in line with the current coding standard.

ksooo commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -2997,12 +2997,12 @@ msgstr ""
 #: system/settings/settings.xml
 msgctxt "#641"
-msgid "Files with ReplayGain information"
+msgid "PreAmp for files with ReplayGain information"
```
Is there  a better word for "PreAmp", which sounds rather technical, or is this something somebody familiar with "ReplayGain" immediately should understand, because it is well-known terminology?

CastagnaIT replied:
different softwares uses "Pre-amp" or "Preamp" or also "Pre-AMP"
but if we want avoid this term a generic idea could be
`Adjustment for files with ReplayGain information`

complexlogic replied:
Preamp is a common concept in ReplayGain, and most ReplayGain-compatible players implement a preamp setting. See the [ReplayGain standard player requirements](https://wiki.hydrogenaud.io/index.php?title=ReplayGain_2.0_specification#Pre-amplification).

ksooo commented on xbmc/application/ApplicationVolumeHandling.h:
```diff
@@ -24,10 +24,10 @@ class CApplicationVolumeHandling
-    int iNoGainPreAmp;
-    int iType;
-    bool bAvoidClipping;
+    float m_preAmp;
+    float m_noGainPreAmp;
+    int m_type;
+    bool m_avoidClipping;
```
We should default initialize the members. See coding guidelines.

ksooo commented:
What about old setting values, stored in existing configurations? Will they still work? Or do they need to be migrated?

ksooo commented on system/settings/settings.xml:
```diff
@@ -271,31 +271,33 @@
-          <control type="spinner" format="string">
-            <formatlabel>14050</formatlabel>
+          <control type="slider" format="number">
+            <popup>true</popup>
```
Do we use this for other settings? I definitely would prefer to see the slider "inline".

CastagnaIT commented:
> What about old setting values, stored in existing configurations? Will they still work? Or do they need to be migrated?

I had to face a similar situation long ago but we dont have a xml settings migration way in kodi
if you dont find other solution to solve the value migration, the only way could be rename the setting names

ksooo commented:
> if you dont find other solution to solve the value migration, the only way could be rename the setting names

Yes, at least this should be done. The way this PR handles the change of settings type semantics by just reusing it is not correct.

ksooo reviewed (changes requested):
Settings "migration" missing or at least introduction of a new setting.

complexlogic commented:
@ksooo @CastagnaIT I just made the following changes:
- Rebased to current `master`
- Changed the GUI control from popup slider to inline slider
- Initialize the `ReplayGainSettings` struct per the coding standard
- Solved the 0 dB sign flip issue by checking the value in `ApplicationVolumeHandling::OnSettingChanged` and setting the value to exact `0.0f` if needed
- Changed the setting ID so it doesn't conflict with the current one

Please let me know if there is any other feedback.

CastagnaIT commented on xbmc/application/ApplicationVolumeHandling.h:
```diff
@@ -23,10 +25,11 @@ class CApplicationVolumeHandling
+    float m_noGainPreAmp;
+    ReplayGain::Type m_type;
+    bool m_avoidClipping;
+    ReplayGainSettings() : m_preAmp(0.0f), m_noGainPreAmp(0.0f), m_type(ReplayGain::NONE), m_avoidClipping(false) {}
```
you need a constructor?

complexlogic replied:
@ksooo  [asked me to add one](https://github.com/xbmc/xbmc/pull/21881#discussion_r980956502) to be aligned with the current coding standard.

CastagnaIT replied:
initialize members not means add a constructor but initialize each single memeber variable

complexlogic replied:
Fixed

CastagnaIT commented on xbmc/settings/Settings.h:
```diff
@@ -274,9 +274,9 @@ class CSettings : public CSettingsBase, public CSettingCreator, public CSettingC
   static constexpr auto SETTING_MUSICPLAYER_SEEKDELAY = "musicplayer.seekdelay";
   static constexpr auto SETTING_MUSICPLAYER_REPLAYGAINTYPE = "musicplayer.replaygaintype";
-  static constexpr auto SETTING_MUSICPLAYER_REPLAYGAINPREAMP = "musicplayer.replaygainpreamp";
+  static constexpr auto SETTING_MUSICPLAYER_REPLAYGAINPREAMP = "musicplayer.replaygainpreamp0";
```
better?
```suggestion
  static constexpr auto SETTING_MUSICPLAYER_REPLAYGAINPREAMP = "musicplayer.replaygainpreampdb";
```

complexlogic replied:
Done

complexlogic committed:
Use PreAmp setting centered at 0, allow fractional dB

fuzzard commented:
@complexlogic is there some straight forward calculation thats possible to convert from the old value to the proposed new values?

Still having the discussion around migration. Whether we need to migrate any existing value to an equivalent (if thats even possible?). Its generally being considered that its bad UX to silently drop a setting without some migration, but im just not sure if such a migration is easily doable.

CrystalP commented:
This was very close to completion. if the only thing left was the setting migration, then yes there is a simple formula
new setting = old setting - 89
see AudioDecoder::373

complexlogic commented:
Does Kodi have a good method of migrating user settings between major versions now? When I created this PR a few years ago, I recall there was not.

CrystalP commented:
no there isn't.

garbear commented:
This looks like a good modernization, if the PR can be reopen and completed.

CrystalP commented:
Hello @complexlogic, there is a framework for setting migration now, are you available to update the PR and take it to the finish line?

complexlogic commented:
@CrystalP Thanks for letting me know. I will try to add setting migration.

CrystalP commented:
I wasn't expecting such a quick answer, that's great! You can resolve the merge conflicts but then I suggest waiting for the merge of #28994 in the next few days. That will make the change in this PR considerably easier. 28994 has the instructions for version bump but I can help, it's very recent code and there isn't much precedent to copy.

CrystalP commented:
28994 is merged.

complexlogic commented:
I rebased this PR to `master`, added the setting migration per #28994, and updated the PR description since this is no longer a breaking change.

garbear commented:
Settings migration added, @ksooo can you approve?

ksooo commented:
Code Formatter is not happy. Please apply requested changes.

ksooo commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -3048,12 +3048,12 @@ msgstr ""
 #: system/settings/settings.xml
 msgctxt "#641"
-msgid "Files with ReplayGain information"
+msgid "Preamp for files with ReplayGain information"
```
Please revert this change. Although I understand the motivation you please should keep in mind, that every string change means that 100+ translations need to be redone for every changed string. Looking at v22 timeline we'd rather lose than win here. The not retranslated strings will only be available in English for the end user. I suggest to open a separate PR for v23 to address this change.

ksooo commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -3048,12 +3048,12 @@ msgstr ""
 #: system/settings/settings.xml
 msgctxt "#642"
-msgid "Files without ReplayGain information"
+msgid "Preamp for files without ReplayGain information"
```
See above. Please revert this change.

ksooo commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -20748,16 +20748,16 @@ msgctxt "#36267"
 #: system/settings/settings.xml
 msgctxt "#36268"
-msgid "Reference volume (PreAmp level) to use for files with encoded ReplayGain information. Default is 89dB as per standard. Change with caution."
+msgid "Preamplification level to use for files with encoded ReplayGain information."
```
See above. Is this change really needed?

ksooo commented on addons/resource.language.en_gb/resources/strings.po:
```diff
@@ -20748,16 +20748,16 @@ msgctxt "#36267"
 #: system/settings/settings.xml
 msgctxt "#36269"
-msgid "Reference volume (PreAmp level) to use for files without encoded ReplayGain information. Default is 89dB as per standard. Change with caution."
+msgid "Preamplification level to use for files without encoded ReplayGain information."
```
See above. Is this change really needed?

ksooo commented on system/settings/settings.xml:
```diff
@@ -391,31 +391,31 @@
           <control type="list" format="string" />
         </setting>
-        <setting id="musicplayer.replaygainpreamp" type="integer" parent="musicplayer.replaygaintype" label="641" help="36268">
+        <setting id="musicplayer.replaygainpreampdb" type="number" parent="musicplayer.replaygaintype" label="641" help="36268">
```
Changing the identifier means users lose their settings in case they have adjusted it. How is migration handled?

Why is the identifier change needed at all?

ksooo commented on system/settings/settings.xml:
```diff
@@ -391,31 +391,31 @@
           </dependencies>
         </setting>
-        <setting id="musicplayer.replaygainnogainpreamp" type="integer" parent="musicplayer.replaygaintype" label="642" help="36269">
+        <setting id="musicplayer.replaygainnogainpreampdb" type="number" parent="musicplayer.replaygaintype" label="642" help="36269">
```
Changing the identifier means users lose their settings in case they have adjusted it. How is migration handled?

Why is the identifier change needed at all?

ksooo commented on xbmc/application/ApplicationVolumeHandling.cpp:
```diff
@@ -185,15 +191,29 @@ bool CApplicationVolumeHandling::OnSettingChanged(const CSetting& setting)
                                      CSettings::SETTING_MUSICPLAYER_REPLAYGAINNOGAINPREAMP))
-    m_replayGainSettings.iNoGainPreAmp = static_cast<const CSettingInt&>(setting).GetValue();
+  {
+    double gain = static_cast<const CSettingNumber&>(setting).GetValue();
```
'const double' please

Why can't this be a 'float'? Later on you cast 'gain' several times to a 'float'. Why not giving it this type from the beginning?

ksooo commented on xbmc/application/ApplicationVolumeHandling.h:
```diff
@@ -29,10 +30,10 @@ class CApplicationVolumeHandling : public IApplicationComponent
+    float m_preAmp = 0.0f;
+    float m_noGainPreAmp = 0.0f;
+    ReplayGain::Type m_type = ReplayGain::NONE;
+    bool m_avoidClipping = false;
```
Use initializers (init value in curly braces), not assignment. Reveals unintended casts in case.

ksooo reviewed (commented):
Looking good. Just nitpicks. And formatting should be fixed before merge.

complexlogic commented:
I applied the formatting patch provided by @jenkins4kodi in https://github.com/xbmc/xbmc/pull/21881/commits/300d511da99e30895b52eb91e183cb07bf73247b. Not sure why it asked me to do it again. When I try to apply the latest formatting patches @jenkins4kodi provided I get the error: `Reversed (or previously applied) patch detected!`.

garbear commented:
You need to fix git history. Commits should be atomic, with correct compiling and formatting between each commit. I'm guessing here we want a single commit. If you squash everything then Jenkins won't complain for interstitial commit formatting.

garbear commented:
There we go, that Jenkins diff should apply cleanly.

garbear commented:
Jenkins build this please.

garbear commented:
Jenkins build this please

garbear commented:
Codex discovered that this is a breaking JSON-RPC API change, so I'll bump the version for you.

AI use:

* **[P2 / blocker] The PR is still externally breaking despite the settings migration.** It removes `musicplayer.replaygainpreamp` and `musicplayer.replaygainnogainpreamp` and replaces them with `...preampdb` IDs.  Kodi's JSON-RPC `Settings.GetSettingValue` / `SetSettingValue` take that ID and directly call `GetSetting(settingId)`; an old ID now produces `InvalidParams`.  The author explicitly said the migration made this “no longer a breaking change,” but it only fixes persisted `settings.xml`, not external consumers.  Also, #28994 explicitly says the migration primitive supports the **same old and new setting name**, so the rename is not required by the framework.  There is a wrinkle: preserving the ID still changes its type/units, so hard-coded API consumers may require an actual compatibility strategy rather than merely retaining the name. At minimum, this PR should be treated/documented as breaking if Kodi considers system setting IDs externally consumable.

The blocker comment I'd put on the PR is:

> The settings migration preserves values in `settings.xml`, but I don't think this makes the change non-breaking. These setting IDs are externally addressable through `Settings.GetSettingValue` / `Settings.SetSettingValue`; after renaming them, clients using the old IDs will get `InvalidParams`.
>
> #28994 says `ConvertSingleSetting()` supports identical old/new setting IDs, so the migration itself doesn't require the rename. If changing the IDs/type is intentional, I think this needs to be called out as a breaking API change; otherwise we need a compatibility path for the old IDs.

garbear commented:
Pushed fixes for two more issues discovered by AI. Fully addresses @ksooo's comment.

AI use:

* **Non-blocking test request:** I'd add one focused migration test because this isn't merely using a fixed prebuilt conversion—the PR supplies its own arithmetic lambda. Test both settings at least at `77 → -12.0`, `89 → 0.0`, and `101 → 12.0`. The migration framework PR specifically says step tests are less worthwhile when only a prebuilt, already-tested converter is used; this one contains custom conversion logic.

* **Tiny nit:** `gain` can still be `const float gain{...};`. @ksooo's latest review asked for const/type cleanup and called the remaining items nitpicks.

garbear commented:
@ksooo Approval when you get a chance is appreciated, shipping B2 soonish.

ksooo reviewed (approved):
LGTM

ksooo commented:
@complexlogic pleaes fulfill one last jenkins format wish and we are ready to merge

CrystalP commented:
I'm not sure that the json version needs bumping. There have been other setting name changes, in v22 and earlier, what about those?
As far as I know setting names are Kodi's internals, and there is no stability guarantee, or is there?
A generic Settings.GetSettingValue / SetSettingValue was a bad idea as far as I am concerned, but since it exists, it's the callers' responsibility to use setting names that match the executing Kodi instance's version.

I don't know if a python api equivalent exists but the same argument would apply.

edit: At least now that setting's versions are tracked there could be a saner way for the caller, since Kodi major version is not granular enough.

garbear commented:
@malard brought up that versioning, especially for JSON-RPC, is messed up, and should be fixed before the B2 release. See https://github.com/xbmc/xbmc/pull/29096. You're right, settings version tracking is probably more appropriate than a JSON-RPC version bump. But I propose we just get this in as-is, and figure out versioning in the next few days.

More in-depth versioning proposal: https://malard.github.io/xbmc/api-version-schedule/

garbear commented:
Android Gradle failure unrelated

garbear commented:
Thanks for the improvement!

complexlogic commented:
Thanks Kodi maintainers for bringing this one back from the dead. Looking forward to the upcoming v22 release.
